### ISSUE TITLE
Is this a bug or a mistake in the comments (or, am I doing it wrong)?

### ISSUE THREAD

**Issue by willemsst** [ReactiveX/RxJava#4543]
I'm using Jersey2 to handle my REST calls so, that gives me the option to handle them async.
Those requests calls are translated in commands placed on a command bus and eventually, a result needs to come back and returned through Jersey. The command bus is constructed with a PublishSubject and so is the result bus where I'm posting the results.

Regularly, I'm seeing timeouts happening in my logs and I narrowed the problem down in the following simple test case. 

``` java
public class TestPublishSubjectBehaviour {

    public static void main(String[] args) {
        Service service = new Service();

        Task task = new Task(1);

        int result = service.publishAndWaitForReply(task)
                .timeout(5, TimeUnit.SECONDS)
                .doOnNext(id -> log("received " + id))
                .doOnError(id -> error("timed out " + task.id))
                .onErrorReturn(e -> 0)
                .doOnSubscribe(() -> log("subscribed results " + task.id))
                .doOnUnsubscribe(() -> log("unsubscribed results " + task.id))
                .toBlocking().first();

        assert result == task.id;
    }

    private static final class Service {

        private final Subject<Integer, Integer> taskResults;

        Service() {
            this.taskResults = PublishSubject.create();
        }

        private static int process(Task task) {
            log("process task " + task.id);
            return task.id;
        }

        Observable<Integer> publishAndWaitForReply(Task task) {
            return this.taskResults
                    .filter(x -> x == task.id)
                    .take(1)
                    .doOnSubscribe(() -> this.taskResults.onNext(
                            process(task)
                    ));
        }
    }

    private static class Task {

        final int id;

        Task(int id) {
            this.id = id;
        }

        @Override
        public boolean equals(Object o) {
            return this == o || !(o == null || getClass() != o.getClass()) && id == ((Task) o).id;

        }

        @Override
        public int hashCode() {
            return id;
        }
    }

    private static void log(String message) {
        System.out.println(LocalDateTime.now() + " - " + Thread.currentThread().getName() + " - " + message);
    }

    private static void error(String message) {
        System.err.println(LocalDateTime.now() + " - " + Thread.currentThread().getName() + " - " + message);
    }
}
```

As you can see, I tried to only publish to the result subject when a subscribe happened, but it looks like the hook happens just before the subscribe instead of after. I think this is a conflict with the documentation. From http://reactivex.io/RxJava/javadoc/rx/Observable.html#doOnSubscribe(rx.functions.Action0) <pre>Modifies the source Observable so that it invokes the given action when it is <strong>subscribed</strong> from its subscribers.</pre>

The expected output of this should be 

```
2016-09-13T15:03:26.767 - main - publish and wait 1
2016-09-13T15:03:27.032 - main - subscribed results 1
2016-09-13T15:03:27.087 - main - process task 1
2016-09-13T15:03:27.088 - main - published result 1
2016-09-13T15:03:27.091 - main - received 1
2016-09-13T15:03:27.093 - main - unsubscribed results 1
```

However, I get

```
2016-09-13T15:37:42.667 - main - subscribed results 1
2016-09-13T15:37:42.691 - main - process task 1
2016-09-13T15:37:47.696 - RxComputationScheduler-1 - timed out 1
```

First question, am I doing it wrong? Is PublishSubject not suited for this? It works if multiple tasks are published though until it gets unsubscribed. 
Alternatively, what are my options? Looking at the other subjects, I think they don't do what I want. I don't want them to hold state for a while. It should just receive and pass on. 

I noticed there's an internal class created to solve a same problem with GroupedObservables, being `BufferUntilSubscriber`

Any help would be much appreciated.

**Comment by akarnokd** [comment]
`doOnSubscribe` happens when the subscription process passes that point in the chain which is before it reaches source `PublishSubject`. At that time, `PublishSubject` has no subscribers so the item is lost.

Why are you using such a convoluted approach? You could just easily use `fromCallable` to compute a result on demand for incoming subscribers?

``` java
Observable<Integer> publishAndWaitForReply(Task task) {
   return Observable.fromCallable(() -> process(task));
}
```

**Comment by willemsst** [comment]
About the lost item, I noticed, hence my question ;)

well.. fromCallable is marked Beta and wasn't in the core when I wrote that code ~20 months ago. I'll check if it works for me as the above example has a very simplified stripped down version of what's going on in the process method.

I would suggest to change the doc line on doOnSubscribe into _is subscribing_ instead of _is subscribed_ as it might be misleading. It was for me.

Also, the solution I came up with isn't _that_ weird, is it? There is one general result stream, subscribe to it with a filter for your id, take 1 and the subscription will get unsubscribed after receiving it because of onCompleted. But too bad one can not postpone putting items on the subject until a subscribe happened.

Thanks for the fast reply.

**Comment by JakeWharton** [comment]
The stable version is `defer(() -> just(process(task)))` or you could just implement `Observable.create`.

The subject and cyclic trigger is very weird to me. It undermines the nature of Rx which is to maintain your state inside of the stream.

**Comment by willemsst** [comment]
To be honest, it has bothered me as well, but I didn't see a solution 'till now. The reasoning came from what I'm using this for.

It's used as the backbone for my CQRS setup. I've a service talking to a command bus -> commands gets handled -> events placed on event bus -> events get projected on read model. And because I wanted to know when the projections were done in the service before I could move on, and in a non-blocking way, I needed to bubble up the 'results' back to the service, so I added two extra subjects where I listen with the mentioned construction. So, I solved it like I described the problem, with 4 different buses: handle stuff, hand over to the next stream when done.

But I got it, I was doing it wrong 😉 Learned something, thx.
